# (PISA Dataset Exploration)
## by (Mohamed Sharif Moawad)

## Preliminary Wrangling

> Briefly introduce your dataset here.

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [3]:
import pandas as pd
import zipfile

zf = zipfile.ZipFile('pisa2012.csv.zip') 
df = pd.read_csv(zf.open('pisa2012.csv'), encoding='latin-1')

/opt/conda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [8]:
df.shape, df.info() ,df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


((485490, 636),
 None,
    Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD       NC  SCHOOLID  \
 0           1  Albania     80000  ALB0006  Non-OECD  Albania         1   
 1           2  Albania     80000  ALB0006  Non-OECD  Albania         1   
 2           3  Albania     80000  ALB0006  Non-OECD  Albania         1   
 3           4  Albania     80000  ALB0006  Non-OECD  Albania         1   
 4           5  Albania     80000  ALB0006  Non-OECD  Albania         1   
 
    STIDSTD  ST01Q01  ST02Q01   ...     W_FSTR75  W_FSTR76 W_FSTR77 W_FSTR78  \
 0        1       10      1.0   ...      13.7954   13.9235  13.1249  13.1249   
 1        2       10      1.0   ...      13.7954   13.9235  13.1249  13.1249   
 2        3        9      1.0   ...      12.7307   12.7307  12.7307  12.7307   
 3        4        9      1.0   ...      12.7307   12.7307  12.7307  12.7307   
 4        5        9      1.0   ...      12.7307   12.7307  12.7307  12.7307   
 
    W_FSTR79 W_FSTR80 WVARSTRR VAR_UNIT SEN

In [10]:
df.describe()

Unnamed: 0      SUBNATIO       SCHOOLID        STIDSTD  \
count  485490.000000  4.854900e+05  485490.000000  485490.000000   
mean   242745.500000  4.315457e+06     240.152197    6134.066201   
std    140149.035432  2.524434e+06     278.563016    6733.144944   
min         1.000000  8.000000e+04       1.000000       1.000000   
25%    121373.250000  2.030000e+06      61.000000    1811.000000   
50%    242745.500000  4.100000e+06     136.000000    3740.000000   
75%    364117.750000  6.880000e+06     291.000000    7456.000000   
max    485490.000000  8.580000e+06    1471.000000   33806.000000   

             ST01Q01        ST02Q01        ST03Q01        ST03Q02  \
count  485490.000000  485438.000000  485490.000000  485490.000000   
mean        9.813323       2.579260       6.558512    1996.070061   
std         3.734726       2.694013       3.705244       0.255250   
min         7.000000       1.000000       1.000000    1996.000000   
25%         9.000000       1.000000       4.000000    1996.000000   
50%        10.000000       1.000000       7.000000    1996.000000   
75%        10.000000       3.000000       9.000000    1996.000000   
max        96.000000      25.000000      99.000000    1997.000000   

             ST06Q01       ST115Q01      ...             W_FSTR74  \
count  457994.000000  479269.000000      ...        485490.000000   
mean        6.148963       1.265356      ...            50.844201   
std         0.970693       0.578992      ...           120.684726   
min         4.000000       1.000000      ...             0.292900   
25%         6.000000       1.000000      ...             4.660300   
50%         6.000000       1.000000      ...            13.637700   
75%         7.000000       1.000000      ...            41.233500   
max        16.000000       4.000000      ...          2476.566800   

            W_FSTR75       W_FSTR76       W_FSTR77       W_FSTR78  \
count  485490.000000  485490.000000  485490.000000  485490.000000   
mean       51.020378      50.943149      50.685275      51.019842   
std       122.946533     121.170883     119.267686     122.981541   
min         0.292900       0.292900       0.292900       0.292900   
25%         4.664800       4.643100       4.667000       4.675200   
50%        13.698900      13.611700      13.672100      13.731100   
75%        41.512500      41.695200      41.097300      41.189600   
max      4155.283000    3743.450100    3232.163700    3904.868100   

            W_FSTR79       W_FSTR80       WVARSTRR       VAR_UNIT  \
count  485490.000000  485490.000000  485490.000000  485490.000000   
mean       50.540724      50.721164      40.013920       1.531189   
std       119.479516     119.799018      22.951264       0.539759   
min         0.292900       0.292900       1.000000       1.000000   
25%         4.651850       4.660300      20.000000       1.000000   
50%        13.582000      13.600200      40.000000       2.000000   
75%        41.290925      41.356000      60.000000       2.000000   
max      3607.478300    3412.174100      80.000000       3.000000   

          SENWGT_STU  
count  485490.000000  
mean        0.140054  
std         0.137864  
min         0.000500  
25%         0.037800  
50%         0.145200  
75%         0.199900  
max         5.095500  

[8 rows x 268 columns]

In [14]:
df['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

###### Some columns have  Massachusetts (USA), Connecticut (USA), and Florida (USA). We have to change that to United States of America.

In [16]:
wrong_cnt=['Massachusetts (USA)','Connecticut (USA)','Florida (USA)']

df['CNT'].replace(wrong_cnt, 'United States of America', inplace=True)

# What is the structure of your dataset?
There are 485490 participants, and there are 636 variables in this dataset. 250 variables are float, 18 are int, and around 368 variables are of type string.
We can get the meaning of the variables meanings by looking them up on the PISA dictionary file.

Here are the most important variables and features:
##### Student Info
CNT: Country
3ST04Q01: Gender
AGE: Age of student

##### Student Math Scores
PV1MATH: Overall Mathematics Score

###### Thoughts
ST29Q02: Worthwhile for Work
ST29Q05: Worthwhile for Career Chances
ST29Q07: Important for Future Study
ST29Q08: Helps to Get a Job

#### Languages
EC05Q01: First language learned
EC06Q01: Age started learning
EC07Q01: Language spoken - Mother
EC07Q02: Language spoken - Father
EC07Q04: Language spoken - Best friend
EC08Q03: Activities language - Internet surfing

###### Math Anxiety
ST42Q01: Worry That It Will Be Difficult
ST42Q05: Get Very Nervous
ST42Q08: Feel Helpless

##### Math Self-Concept
ST42Q02: Not Good at Maths
ST42Q04: Get Good (Grades)
ST42Q06: Learn Quickly
ST29Q04: Enjoy Maths


###### Math Self-Efficacy
ST37Q01	Math Self-Efficacy - Using a Train Timetable
ST37Q02	Math Self-Efficacy - Calculating TV Discount
ST37Q03	Math Self-Efficacy - Calculating Square Metres of Tiles
ST37Q06	Math Self-Efficacy - Distance to Scale
ST37Q07	Math Self-Efficacy - Solving Equation 2
ST37Q08	Math Self-Efficacy - Calculate Petrol Consumption Rate


##### Math Behaviour
ST49Q02: Help Friends with Maths
ST49Q04: Participate in Competitions
ST49Q05: Study More Than 2 Extra Hours a Day
ST49Q06: Play Chess
ST49Q07: Computer programming
ST49Q09: Participate in Math Club
MATBEH   Mathematics Behaviour
MATHEFF  Mathematics Self-Efficacy
MATWKETH Mathematics Work Ethic
ST46Q02: Work Hard on Homework
ST46Q03: Prepared for Exams
ST46Q09: Keep Work Organized

##### Attitude 
ATSCHL: Attitude towards School: Learning Outcomes
ATTLNACT: Attitude towards School: Learning Activities
BELONG: Sense of Belonging to School

##### Subjective Norms
ST35Q04: Parents Believe Studying Mathematics Is Important
ST35Q05: Parents Believe Mathematics Is Important for Career
ST35Q06: Parents Like Mathematics

##### Parents
BFMJ2: Father SQ ISEI
BMMJ1: Mother SQ ISEI
WEALTH	Wealth
MISCED	Educational level of mother (ISCED)
OCOD1	ISCO-08 Occupation code - Mother
OCOD2	ISCO-08 Occupation code - Father
PARED	Highest parental education in years

##### Math Classroom
ANCCLSMAN	Mathematics Teacher's Classroom Management (Anchored)
ANCCOGACT	Cognitive Activation in Mathematics Lessons (Anchored)
ANCINSTMOT	Instrumental Motivation for Mathematics (Anchored)
ANCINTMAT	Mathematics Interest (Anchored)
ANCMATWKETH	Mathematics Work Ethic (Anchored)
ANCMTSUP	Mathematics Teacher's Support (Anchored)
ANCSCMAT	Mathematics Self-Concept (Anchored)
ANCSTUDREL	Teacher Student Relations (Anchored)
USEMATH 	Use of ICT in Mathematic Lessons

##### Teacher Behavior
TCHBEHFA	Teacher Behaviour: Formative Assessment
TCHBEHSO	Teacher Behaviour: Student Orientation
TCHBEHTD	Teacher Behaviour: Teacher-directed Instruction
TEACHSUP	Teacher Support

#### Experience
ST61Q01	Experience with Applied Maths Tasks - Use <Train Timetable>
ST61Q02	Experience with Applied Maths Tasks - Calculate Price including Tax
ST61Q03	Experience with Applied Maths Tasks - Calculate Square Metres
ST61Q04	Experience with Applied Maths Tasks - Understand Scientific Tables
ST61Q05	Experience with Pure Maths Tasks - Solve Equation 1
ST61Q06	Experience with Applied Maths Tasks - Use a Map to Calculate Distance
ST61Q07	Experience with Pure Maths Tasks - Solve Equation 2
ST61Q08	Experience with Applied Maths Tasks - Calculate Power Consumption Rate
ST61Q09	Experience with Applied Maths Tasks - Solve Equation 3
ST62Q01	Familiarity with Math Concepts - Exponential Function
ST62Q02	Familiarity with Math Concepts - Divisor
ST62Q03	Familiarity with Math Concepts - Quadratic Function
ST62Q04	Overclaiming - Proper Number
ST62Q06	Familiarity with Math Concepts - Linear Equation
ST62Q07	Familiarity with Math Concepts - Vectors
ST62Q08	Familiarity with Math Concepts - Complex Number
ST62Q09	Familiarity with Math Concepts - Rational Number
ST62Q10	Familiarity with Math Concepts - Radicals
ST62Q11	Overclaiming - Subjunctive Scaling
ST62Q12	Familiarity with Math Concepts - Polygon
ST62Q13	Overclaiming - Declarative Fraction
ST62Q15	Familiarity with Math Concepts - Congruent Figure
ST62Q16	Familiarity with Math Concepts - Cosine
ST62Q17	Familiarity with Math Concepts - Arithmetic Mean
ST62Q19	Familiarity with Math Concepts - Probability

#### Future Orientation
EC03Q01 	Future Orientation - Internship
EC03Q02	    Future Orientation - Work-site visits
EC03Q03	    Future Orientation - Job fair
EC03Q04 	Future Orientation - Career advisor at school
EC03Q05	    Future Orientation - Career advisor outside school
EC03Q06	    Future Orientation - Questionnaire
EC03Q07	    Future Orientation - Internet search
EC03Q08	    Future Orientation - Tour<ISCED 3-5> institution
EC03Q09	    Future Orientation - web search <ISCED 3-5> prog
EC03Q10	    Future Orientation - country specific item

#### Math Lessons
IC11Q01	    Maths lessons - Draw graph
IC11Q02	    Maths lessons - Calculation with numbers
IC11Q03	    Maths lessons - Geometric figures
IC11Q04	    Maths lessons - Spreadsheet
IC11Q05	    Maths lessons - Algebra
IC11Q06	    Maths lessons - Histograms
IC11Q07	    Maths lessons - Change in graphs
#### Attitudes
IC22Q01	    Attitudes - Useful for schoolwork
IC22Q02	    Attitudes - Homework more fun
IC22Q04	    Attitudes - Source of information
IC22Q06	    Attitudes - Troublesome
IC22Q07	    Attitudes - Not suitable for schoolwork
IC22Q08	    Attitudes - Too unreliable
MMINS	    Learning time (minutes per week)
IC03Q01	    First use of computers
IC04Q01	    First access to Internet
#### At School
IC10Q01	At School - Chat on line
IC10Q02	At School - Email
IC10Q03	At School - Browse for schoolwork
IC10Q04	At School - Download from website
IC10Q05	At School - Post on website
IC10Q06	At School - Simulations
IC10Q07	At School - Practice and drilling
IC10Q08	At School - Homework
IC10Q09	At School - Group work
IC02Q01	At school - Desktop Computer
IC02Q02	At school - Portable laptop
IC02Q03	At school - Tablet computer
IC02Q04	At school - Internet connection
IC02Q05	At school - Printer
IC02Q06	At school - USB (memory) stick
IC02Q07	At school - Ebook reader
#### Out of School
IC08Q01	Out-of-school 8 - One player games.
IC08Q02	Out-of-school 8 - ColLabourative games.
IC08Q03	Out-of-school 8 - Use email
IC08Q04	Out-of-school 8 - Chat on line
IC08Q05	Out-of-school 8 - Social networks
IC08Q06	Out-of-school 8 - Browse the Internet for fun
IC08Q07	Out-of-school 8 - Read news
IC08Q08	Out-of-school 8 - Obtain practical information from the Internet
IC08Q09	Out-of-school 8 - Download music
IC08Q11	Out-of-school 8 - Upload content
IC09Q01	Out-of-school 9 - Internet for school
IC09Q02	Out-of-school 9 - Email students
IC09Q03	Out-of-school 9 - Email teachers
IC09Q04	Out-of-school 9 - Download from School
IC09Q05	Out-of-school 9 - Announcements
IC09Q06	Out-of-school 9 - Homework
IC09Q07	Out-of-school 9 - Share school material

# What is/are the main feature(s) of interest in your dataset?
I am most interested in figuring out what features are best for predicting the academic performance and acheivemnts.

# What features in the dataset do you think will help support your investigation into your feature(s) of interest?
I expect that location and student attitudes would play a role in identifyinig the academic acheivement of the student. I would also test the dataset to see if there is a difference in academic performance based on gender or inequality.

In [2]:
### Categorize the variables
student_info=['CNT', '3ST04Q01','AGE']

##### Student Math Scores
maths_score=['PV1MATH']

###### Thoughts
thought=['ST29Q02', 'ST29Q05', 'ST29Q07', 'ST29Q08']

#### Languages
Languages=['EC05Q01','EC06Q01','EC07Q01','EC08Q03']

###### Math Anxiety
Anxious=['ST42Q01', 'ST42Q05','ST42Q08']

##### Math Self-Concept
love_maths=['ST42Q02','ST42Q04', 'ST42Q06','ST29Q04']


###### Math Self-Efficacy
math_efficacy=['ST37Q01','ST37Q02','ST37Q03','ST37Q06','ST37Q07','ST37Q08']

##### Math Behaviour
behaviour=['ST49Q02','ST49Q04','ST49Q05','ST49Q06','ST49Q07','ST49Q09']
        
metrics= ['MATBEH','MATHEFF','MATWKETH']
exp= ['ST46Q02','ST46Q03','ST46Q09']

##### Attitude 
attitude=['ATSCHL','ATTLNACT','BELONG']


##### Subjective Norms
par_thought=['ST35Q04','ST35Q05','ST35Q06']

##### Parents
parents=['BFMJ2','BMMJ1','WEALTH', 'MISCED', 'OCOD1','OCOD2', 'PARED']

##### Math Classroom
maths_class=['ANCCOGACT','ANCINSTMOT','ANCSTUDREL','USEMATH']

##### Teacher Behavior
teacher=['TCHBEHFA','TCHBEHSO','TCHBEHTD','TEACHSUP']

#### Experience
math_exp=['ST61Q01','ST61Q04','ST61Q05']


#### Future Orientation
future=['EC03Q01','EC03Q03','EC03Q05','EC03Q10']

#### Math Lessons
math_less=['IC11Q03','IC11Q04','IC11Q05']
#### Attitudes
attitudes=['IC22Q01','MMINS','IC03Q01','IC04Q01']

In [4]:
variables=[thought,Languages,Anxious,love_maths,math_efficacy,behaviour,metrics,exp,attitude,par_thought,parents,
          maths_class,teacher,math_exp,future,math_less,attitudes]

In [37]:
i=0
for _ in variables:
    i+=1
    for feature in _:
        print(i,'\n',df[feature].value_counts(),'\n')

1 
 Agree                150479
Strongly agree        97133
Disagree              49792
Strongly disagree     18069
Name: ST29Q02, dtype: int64 

1 
 Agree                153313
Strongly agree        99425
Disagree              44218
Strongly disagree     18275
Name: ST29Q05, dtype: int64 

1 
 Agree                132131
Strongly agree        90470
Disagree              65853
Strongly disagree     26612
Name: ST29Q07, dtype: int64 

1 
 Agree                150347
Strongly agree        85236
Disagree              57949
Strongly disagree     21700
Name: ST29Q08, dtype: int64 

2 
 <test language> or <other official national language(s) or d    119401
other than <test language> or <other official national langu     10257
Name: EC05Q01, dtype: int64 

2 
 0 to 3 years         26070
4 to 6 years          9207
7 to 9 years          2994
10 to 12 years        1275
13 years or older      799
Name: EC06Q01, dtype: int64 

2 
 Mostly <heritage language>                               23093
Most

10 
 Agree                144854
Strongly agree       115112
Disagree              46278
Strongly disagree      8599
Name: ST35Q05, dtype: int64 

10 
 Agree                148876
Disagree              93913
Strongly agree        49488
Strongly disagree     21112
Name: ST35Q06, dtype: int64 

11 
 25.95    11971
22.57    10638
30.34     9953
11.74     9425
30.78     8549
51.50     7166
70.34     6915
33.76     6670
35.34     6648
23.80     5375
43.33     5369
71.39     5280
51.56     5094
65.01     4793
11.01     4572
27.91     4549
82.41     4330
26.62     4171
15.35     4115
36.35     4109
28.48     3996
24.53     3978
26.85     3730
76.65     3575
59.89     3524
79.74     3498
11.56     3407
28.52     3371
57.03     3312
29.16     3283
         ...  
44.87       56
52.57       49
51.77       47
34.12       47
28.08       47
54.67       44
17.85       42
83.81       42
48.27       40
56.98       40
74.50       39
23.97       39
25.49       38
65.23       37
22.40       35
84.16      

14 
 Sometimes     125923
Rarely         80428
Frequently     53983
Never          52465
Name: ST61Q01, dtype: int64 

14 
 Sometimes     108115
Rarely         93437
Never          59531
Frequently     49221
Name: ST61Q04, dtype: int64 

14 
 Frequently    189647
Sometimes      76491
Rarely         27224
Never          18336
Name: ST61Q05, dtype: int64 

15 
 No, never    120999
Yes           39534
Name: EC03Q01, dtype: int64 

15 
 No, never    118784
Yes           46413
Name: EC03Q03, dtype: int64 

15 
 No, never    133874
Yes           31110
Name: EC03Q05, dtype: int64 

15 
 No, never    69596
Yes          62943
Name: EC03Q10, dtype: int64 

16 
 No                                              203795
Yes, students did this                           44102
Yes, but only the teacher demonstrated this      40971
Name: IC11Q03, dtype: int64 

16 
 No                                              194896
Yes, students did this                           57750
Yes, but only the teacher demo

In [38]:
necessary_data=['CNT', 'ST04Q01','AGE','PV1MATH','ST29Q02', 'ST29Q05', 'ST29Q07',
                'ST29Q08','EC05Q01','EC06Q01','EC07Q01','EC08Q03','ST42Q01', 'ST42Q05','ST42Q08',
                'ST42Q02','ST42Q04', 'ST42Q06','ST29Q04','ST37Q01','ST37Q02','ST37Q03',
                'ST37Q06','ST37Q07','ST37Q08','ST49Q02','ST49Q04','ST49Q05','ST49Q06','ST49Q07',
                'ST49Q09','MATBEH','MATHEFF','MATWKETH','ST46Q02','ST46Q03','ST46Q09',
                'ATSCHL','ATTLNACT','BELONG','ST35Q04','ST35Q05','ST35Q06',
                'BFMJ2','BMMJ1','WEALTH', 'MISCED', 'OCOD1','OCOD2', 'PARED',
                'ANCCOGACT','ANCINSTMOT','ANCMTSUP','ANCSTUDREL','USEMATH',
                'TCHBEHFA','TCHBEHSO','TCHBEHTD','TEACHSUP',
                'ST61Q01','ST61Q04','ST61Q05','EC03Q01','EC03Q03',
                'EC03Q05','EC03Q10','IC11Q03','IC11Q04','IC11Q05',
                'IC22Q01','MMINS','IC03Q01','IC04Q01']

In [39]:
df_new=df[necessary_data]

In [40]:
df_new.head()

CNT ST04Q01    AGE   PV1MATH         ST29Q02         ST29Q05  \
0  Albania  Female  16.17  406.8469  Strongly agree           Agree   
1  Albania  Female  16.17  486.1427  Strongly agree           Agree   
2  Albania  Female  15.58  533.2684  Strongly agree  Strongly agree   
3  Albania  Female  15.67  412.2215             NaN             NaN   
4  Albania  Female  15.50  381.9209  Strongly agree  Strongly agree   

          ST29Q07         ST29Q08 EC05Q01 EC06Q01   ...   EC03Q03 EC03Q05  \
0           Agree  Strongly agree     NaN     NaN   ...       NaN     NaN   
1        Disagree        Disagree     NaN     NaN   ...       NaN     NaN   
2  Strongly agree  Strongly agree     NaN     NaN   ...       NaN     NaN   
3             NaN             NaN     NaN     NaN   ...       NaN     NaN   
4  Strongly agree  Strongly agree     NaN     NaN   ...       NaN     NaN   

  EC03Q10 IC11Q03 IC11Q04 IC11Q05 IC22Q01  MMINS IC03Q01 IC04Q01  
0     NaN     NaN     NaN     NaN     NaN    NaN     NaN     NaN  
1     NaN     NaN     NaN     NaN     NaN  270.0     NaN     NaN  
2     NaN     NaN     NaN     NaN     NaN    NaN     NaN     NaN  
3     NaN     NaN     NaN     NaN     NaN  135.0     NaN     NaN  
4     NaN     NaN     NaN     NaN     NaN    NaN     NaN     NaN  

[5 rows x 73 columns]

In [41]:
#Calculates the number of null values for each section 
#as well as percent of students who answered all questions under each section
i=0
for _ in variables:
    i+=1
    print(i)
    d=dict(df_new[_].isnull().sum(axis=1).value_counts())
    print(d)
    print('% Complete: {}'.format(d[0]/485490))
    print('\n')

1
{0: 312694, 4: 168774, 1: 2822, 3: 640, 2: 560}
% Complete: 0.6440791777379554


2
{4: 351716, 3: 86979, 0: 33199, 1: 10537, 2: 3059}
% Complete: 0.06838245895898988


3
{0: 310240, 3: 170715, 1: 3680, 2: 855}
% Complete: 0.6390244907207151


4
{3: 159771, 1: 158133, 0: 153294, 4: 11692, 2: 2600}
% Complete: 0.3157510968300068


5
{0: 310166, 6: 169512, 1: 3847, 5: 737, 2: 615, 3: 357, 4: 256}
% Complete: 0.6388720673958269


6
{0: 308751, 6: 171409, 1: 3816, 2: 665, 5: 429, 3: 218, 4: 202}
% Complete: 0.6359574862510041


7
{0: 311099, 3: 167682, 1: 4290, 2: 2419}
% Complete: 0.6407938371542153


8
{0: 311052, 3: 170888, 1: 2895, 2: 655}
% Complete: 0.6406970277451647


9
{0: 310031, 3: 170808, 1: 2914, 2: 1737}
% Complete: 0.6385939978166388


10
{0: 311682, 3: 169212, 1: 3750, 2: 846}
% Complete: 0.6419946857813755


11
{0: 320692, 1: 121969, 2: 31494, 3: 4367, 5: 4116, 4: 2566, 7: 252, 6: 34}
% Complete: 0.6605532554738511


12
{1: 154980, 3: 101293, 0: 94199, 4: 75688, 2: 59330}

We should drop 12 & 2 due to that its majority is null values and therfore would provide inconsistent data.

In [42]:
df_new.drop(columns=['EC05Q01','EC06Q01','EC07Q01','EC08Q03','ANCCOGACT','ANCINSTMOT','ANCSTUDREL','USEMATH'],inplace=True)

/opt/conda/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [46]:
df_new.head()

CNT ST04Q01    AGE   PV1MATH         ST29Q02         ST29Q05  \
0  Albania  Female  16.17  406.8469  Strongly agree           Agree   
1  Albania  Female  16.17  486.1427  Strongly agree           Agree   
2  Albania  Female  15.58  533.2684  Strongly agree  Strongly agree   
3  Albania  Female  15.67  412.2215             NaN             NaN   
4  Albania  Female  15.50  381.9209  Strongly agree  Strongly agree   

          ST29Q07         ST29Q08         ST42Q01         ST42Q05   ...    \
0           Agree  Strongly agree           Agree           Agree   ...     
1        Disagree        Disagree             NaN             NaN   ...     
2  Strongly agree  Strongly agree             NaN             NaN   ...     
3             NaN             NaN             NaN           Agree   ...     
4  Strongly agree  Strongly agree  Strongly agree  Strongly agree   ...     

  EC03Q03 EC03Q05 EC03Q10 IC11Q03 IC11Q04 IC11Q05 IC22Q01  MMINS IC03Q01  \
0     NaN     NaN     NaN     NaN     NaN     NaN     NaN    NaN     NaN   
1     NaN     NaN     NaN     NaN     NaN     NaN     NaN  270.0     NaN   
2     NaN     NaN     NaN     NaN     NaN     NaN     NaN    NaN     NaN   
3     NaN     NaN     NaN     NaN     NaN     NaN     NaN  135.0     NaN   
4     NaN     NaN     NaN     NaN     NaN     NaN     NaN    NaN     NaN   

  IC04Q01  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN  

[5 rows x 65 columns]

In [48]:
#agreement were available at 1,3,4,8,10 with reverse agreement on the 'ST42Q02' of the fourth category
for _ in ['ST29Q05', 'ST29Q07', 'ST29Q08','ST42Q01', 'ST42Q05','ST42Q08','ST42Q04', 'ST42Q06','ST29Q04'
          ,'ST29Q02','ST46Q02','ST46Q03','ST46Q09','ST35Q04','ST35Q05','ST35Q06']:
    
    df_new.loc[:,_]=df_new.loc[:,_].map({'Strongly agree':4,'Agree': 3,'Disagree': 2, 'Strongly disagree':1})

/opt/conda/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [49]:
df_new.loc[:,'ST42Q02']=df_new.loc[:,'ST42Q02'].map({'Strongly agree':1, 
                                             'Agree': 2,'Disagree': 3, 'Strongly disagree':4})

/opt/conda/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [52]:
# The sixth category had different system
for _ in ['ST49Q02','ST49Q04','ST49Q05','ST49Q06','ST49Q07','ST49Q09']:
    df_new.loc[:,_]=df_new.loc[:,_].map({'Always or almost always': 4, 'Often':3, 'Sometimes':2, 'Never or rarely':1})

/opt/conda/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [53]:
for _ in ['ST37Q01','ST37Q02','ST37Q03','ST37Q06','ST37Q07','ST37Q08']:
    df_new.loc[:,_]=df_new.loc[:,_].map({'Confident': 4, 'Not very confident':3, 'Very confident':2, 'Not at all confident':1})

/opt/conda/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [55]:
df_new.head()

CNT ST04Q01    AGE   PV1MATH  ST29Q02  ST29Q05  ST29Q07  ST29Q08  \
0  Albania  Female  16.17  406.8469      4.0      3.0      3.0      4.0   
1  Albania  Female  16.17  486.1427      4.0      3.0      2.0      2.0   
2  Albania  Female  15.58  533.2684      4.0      4.0      4.0      4.0   
3  Albania  Female  15.67  412.2215      NaN      NaN      NaN      NaN   
4  Albania  Female  15.50  381.9209      4.0      4.0      4.0      4.0   

   ST42Q01  ST42Q05   ...     EC03Q03  EC03Q05  EC03Q10  IC11Q03  IC11Q04  \
0      3.0      3.0   ...         NaN      NaN      NaN      NaN      NaN   
1      NaN      NaN   ...         NaN      NaN      NaN      NaN      NaN   
2      NaN      NaN   ...         NaN      NaN      NaN      NaN      NaN   
3      NaN      3.0   ...         NaN      NaN      NaN      NaN      NaN   
4      4.0      4.0   ...         NaN      NaN      NaN      NaN      NaN   

   IC11Q05  IC22Q01  MMINS  IC03Q01  IC04Q01  
0      NaN      NaN    NaN      NaN      NaN  
1      NaN      NaN  270.0      NaN      NaN  
2      NaN      NaN    NaN      NaN      NaN  
3      NaN      NaN  135.0      NaN      NaN  
4      NaN      NaN    NaN      NaN      NaN  

[5 rows x 65 columns]

we are going to get the average of all columns since they are all of a 4-scale system. Then, we are going to sum the different variables of the same category to reduce the number of columns. and we would drop any column with a full null category

In [56]:
df_new['math_efficacy']=df_new.loc[:,math_efficacy].sum(axis=1, skipna=False)/len(math_efficacy)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [57]:
df_new['thought']=df_new.loc[:,thought].sum(axis=1, skipna=False)/len(thought)
df_new['Anxious']=df_new.loc[:,Anxious].sum(axis=1, skipna=False)/len(Anxious)
df_new['love_maths']=df_new.loc[:,love_maths].sum(axis=1, skipna=False)/len(love_maths)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/p

In [58]:
df_new['exp']=df_new.loc[:,exp].sum(axis=1, skipna=False)/len(exp)
df_new['par_thought']=df_new.loc[:,par_thought].sum(axis=1, skipna=False)/len(par_thought)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [59]:
df_new['behaviour']=df_new.loc[:,behaviour].sum(axis=1, skipna=False)/len(behaviour)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [61]:
df_new.head()

CNT ST04Q01    AGE   PV1MATH  ST29Q02  ST29Q05  ST29Q07  ST29Q08  \
0  Albania  Female  16.17  406.8469      4.0      3.0      3.0      4.0   
1  Albania  Female  16.17  486.1427      4.0      3.0      2.0      2.0   
2  Albania  Female  15.58  533.2684      4.0      4.0      4.0      4.0   
3  Albania  Female  15.67  412.2215      NaN      NaN      NaN      NaN   
4  Albania  Female  15.50  381.9209      4.0      4.0      4.0      4.0   

   ST42Q01  ST42Q05    ...      MMINS  IC03Q01  IC04Q01  math_efficacy  \
0      3.0      3.0    ...        NaN      NaN      NaN       2.500000   
1      NaN      NaN    ...      270.0      NaN      NaN       2.833333   
2      NaN      NaN    ...        NaN      NaN      NaN       2.833333   
3      NaN      3.0    ...      135.0      NaN      NaN            NaN   
4      4.0      4.0    ...        NaN      NaN      NaN            NaN   

   thought   Anxious  love_maths       exp  par_thought  behaviour  
0     3.50  2.666667        3.00  3.000000     3.000000        NaN  
1     2.75       NaN         NaN  3.333333     2.666667        NaN  
2     4.00       NaN         NaN  4.000000     3.666667        NaN  
3      NaN       NaN         NaN       NaN          NaN        NaN  
4     4.00  3.333333        2.25  3.666667     3.666667        NaN  

[5 rows x 72 columns]

In [64]:
df_new.drop(columns=math_efficacy+thought+Anxious+love_maths+exp+par_thought+behaviour, axis=1,inplace=True)

KeyError: "['ST37Q01' 'ST37Q02' 'ST37Q03' 'ST37Q06' 'ST37Q07' 'ST37Q08' 'ST29Q02'\n 'ST29Q05' 'ST29Q07' 'ST29Q08' 'ST42Q01' 'ST42Q05' 'ST42Q08' 'ST42Q02'\n 'ST42Q04' 'ST42Q06' 'ST29Q04' 'ST46Q02' 'ST46Q03' 'ST46Q09' 'ST35Q04'\n 'ST35Q05' 'ST35Q06' 'ST49Q02' 'ST49Q04' 'ST49Q05' 'ST49Q06' 'ST49Q07'\n 'ST49Q09'] not found in axis"

In [65]:
df_new.head()

CNT ST04Q01    AGE   PV1MATH  MATBEH  MATHEFF  MATWKETH  ATSCHL  \
0  Albania  Female  16.17  406.8469  0.6426    -0.77    0.2882   -2.31   
1  Albania  Female  16.17  486.1427  1.4702     0.34    0.6490     NaN   
2  Albania  Female  15.58  533.2684  0.9618     0.34    2.0389     NaN   
3  Albania  Female  15.67  412.2215     NaN      NaN       NaN     NaN   
4  Albania  Female  15.50  381.9209  1.8169     0.41    1.6881    1.38   

   ATTLNACT  BELONG    ...      MMINS  IC03Q01  IC04Q01 math_efficacy thought  \
0    0.5206   -1.18    ...        NaN      NaN      NaN      2.500000    3.50   
1       NaN     NaN    ...      270.0      NaN      NaN      2.833333    2.75   
2       NaN     NaN    ...        NaN      NaN      NaN      2.833333    4.00   
3       NaN     NaN    ...      135.0      NaN      NaN           NaN     NaN   
4    1.2115    2.63    ...        NaN      NaN      NaN           NaN    4.00   

    Anxious  love_maths       exp  par_thought  behaviour  
0  2.666667        3.00  3.000000     3.000000        NaN  
1       NaN         NaN  3.333333     2.666667        NaN  
2       NaN         NaN  4.000000     3.666667        NaN  
3       NaN         NaN       NaN          NaN        NaN  
4  3.333333        2.25  3.666667     3.666667        NaN  

[5 rows x 43 columns]

In [68]:
column_names={'ST04Q01':'gender', 'PV1MATH':'score'}
df_new.rename(columns=column_names,inplace=True)

/opt/conda/lib/python3.6/site-packages/pandas/core/frame.py:3781: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


In [69]:
df_new.to_csv('data_cleaned.csv', index_label=False)

In [3]:
df_new=pd.read_csv('data_cleaned.csv')

/opt/conda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (26,27,28,29,30,31,32,33,35,36) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


#### After doing that, I found out some data were left off without cleaninng. I would start from here and clean on the preliminary cleaned data to avoid uploading the big zip file

In [9]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 485490 entries, 0 to 485489
Data columns (total 43 columns):
CNT              485490 non-null object
gender           485490 non-null object
AGE              485374 non-null float64
score            485490 non-null float64
MATBEH           313847 non-null float64
MATHEFF          315948 non-null float64
MATWKETH         314501 non-null float64
ATSCHL           312584 non-null float64
ATTLNACT         311675 non-null float64
BELONG           313399 non-null float64
BFMJ2            416150 non-null float64
BMMJ1            364814 non-null float64
WEALTH           479597 non-null float64
MISCED           467085 non-null object
OCOD1            483887 non-null object
OCOD2            482936 non-null object
PARED            473091 non-null float64
ANCMTSUP         308631 non-null float64
TCHBEHFA         314678 non-null float64
TCHBEHSO         315114 non-null float64
TCHBEHTD         315519 non-null float64
TEACHSUP         316371 non-null 

In [5]:
variables=[math_exp,future,math_less,attitudes]

In [11]:
i=0
for _ in variables:
    i+=1
    for feature in _:
        print(i,'\n',df_new[feature].value_counts(),'\n')

1 
 Sometimes     125923
Rarely         80428
Frequently     53983
Never          52465
Name: ST61Q01, dtype: int64 

1 
 Sometimes     108115
Rarely         93437
Never          59531
Frequently     49221
Name: ST61Q04, dtype: int64 

1 
 Frequently    189647
Sometimes      76491
Rarely         27224
Never          18336
Name: ST61Q05, dtype: int64 

2 
 No, never    120999
Yes           39534
Name: EC03Q01, dtype: int64 

2 
 No, never    118784
Yes           46413
Name: EC03Q03, dtype: int64 

2 
 No, never    133874
Yes           31110
Name: EC03Q05, dtype: int64 

2 
 No, never    69596
Yes          62943
Name: EC03Q10, dtype: int64 

3 
 No                                              203795
Yes, students did this                           44102
Yes, but only the teacher demonstrated this      40971
Name: IC11Q03, dtype: int64 

3 
 No                                              194896
Yes, students did this                           57750
Yes, but only the teacher demonstrated 

In [6]:
for _ in ['ST61Q01','ST61Q04','ST61Q05']:
    df_new.loc[:,_]=df_new.loc[:,_].map({'Frequently': 4, 'Sometimes':3, 'Rarely':2, 'Never':1})

In [7]:
df_new['math_exp']=df_new.loc[:,math_exp].sum(axis=1, skipna=False)/len(math_exp)

In [8]:
for _ in ['IC11Q03','IC11Q04','IC11Q05']:
    df_new.loc[:,_]=df_new.loc[:,_].map({'No': 0, 'Yes, students did this':1, 'Yes, but only the teacher demonstrated this':2})

In [9]:
df_new['math_lesson']=df_new.loc[:,math_less].sum(axis=1, skipna=False)/len(math_less)

In [10]:
df_new.drop(columns=math_less+math_exp, axis=1,inplace=True)

In [11]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 485490 entries, 0 to 485489
Data columns (total 39 columns):
CNT              485490 non-null object
gender           485490 non-null object
AGE              485374 non-null float64
score            485490 non-null float64
MATBEH           313847 non-null float64
MATHEFF          315948 non-null float64
MATWKETH         314501 non-null float64
ATSCHL           312584 non-null float64
ATTLNACT         311675 non-null float64
BELONG           313399 non-null float64
BFMJ2            416150 non-null float64
BMMJ1            364814 non-null float64
WEALTH           479597 non-null float64
MISCED           467085 non-null object
OCOD1            483887 non-null object
OCOD2            482936 non-null object
PARED            473091 non-null float64
ANCMTSUP         308631 non-null float64
TCHBEHFA         314678 non-null float64
TCHBEHSO         315114 non-null float64
TCHBEHTD         315519 non-null float64
TEACHSUP         316371 non-null 

### I think some variables won't be of much value so I would delete them:
##### ANCMTSUP	Mathematics Teacher's Support (Anchored), we would use TEACHSUP only.
##### TCHBEHSO	Teacher Behaviour: Student Orientation, not useful
##### TCHBEHTD	Teacher Behaviour: Teacher-directed Instruction, not useful
##### TCHBEHFA	Teacher Behaviour: Formative Assessment, not useful
###### behaviour   All values are null.
###### EC03Q03	         Future Orientation - Job fair, not useful
###### EC03Q05	         Future Orientation - Career advisor outside school, not useful
###### EC03Q10	         Future Orientation - country specific item, not useful
###### IC22Q01          Attitudes - Useful for schoolwork, some metrics would reflect on that so we think there is no need for that.



In [12]:
df_new.drop(columns=['ANCMTSUP','TCHBEHSO','TCHBEHTD','TCHBEHFA','behaviour','EC03Q03','EC03Q05','EC03Q10'
                    ,'IC22Q01'], axis=1,inplace=True)

In [14]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 485490 entries, 0 to 485489
Data columns (total 30 columns):
CNT              485490 non-null object
gender           485490 non-null object
AGE              485374 non-null float64
score            485490 non-null float64
MATBEH           313847 non-null float64
MATHEFF          315948 non-null float64
MATWKETH         314501 non-null float64
ATSCHL           312584 non-null float64
ATTLNACT         311675 non-null float64
BELONG           313399 non-null float64
BFMJ2            416150 non-null float64
BMMJ1            364814 non-null float64
WEALTH           479597 non-null float64
MISCED           467085 non-null object
OCOD1            483887 non-null object
OCOD2            482936 non-null object
PARED            473091 non-null float64
TEACHSUP         316371 non-null float64
EC03Q01          160533 non-null object
MMINS            283303 non-null float64
IC03Q01          293216 non-null object
IC04Q01          296305 non-null ob

In [17]:
column_names={'MATBEH':'math_behav','MATHEFF':'math_self_efficacy','MATWKETH':'math_work_ethic','ATSCHL':'att_learn_otcm',
              'ATTLNACT':'att_learn_actv',
              'BFMJ2':'f_sq',
              'BMMJ1':'m_sq',
              'OCOD1':'m_occ',
              'OCOD2':'f_occ',
              'PARED':'parent_ed',
              'TEACHSUP':'teach_supp',
              'EC03Q01':'internship',
              'MMINS':'Learning time',
              'IC03Q01':'comp_frst',
              'IC04Q01':'intrnt_frst'}      
df_new.rename(columns=column_names,inplace=True)

In [18]:
df_new.to_csv('data_cleaned.csv', index_label=False)